### DISCUSSION TITLE
atomic fingerprints

### DISCUSSION THREAD

**Discussion by ekwan** [rdkit/rdkit#4268] [Q&A]
Hello all,

Is it possible to get any of the intermediate hash codes from a fingerprint calculation?  Reading [this description](https://towardsdatascience.com/a-practical-introduction-to-the-use-of-molecular-fingerprints-in-drug-discovery-7f15021be2b1) of the fingerprinting procedure indicates that hash codes are generated at each atom before being folded into an overall molecular fingerprint.

Also, is it possible to compute these fingerprints from a 3D structure, rather than a SMILES string?

Thanks,
-Eugene

**Comment by greglandrum** [comment]
Hi Eugene,

> 
> Is it possible to get any of the intermediate hash codes from a fingerprint calculation? Reading [this description](https://towardsdatascience.com/a-practical-introduction-to-the-use-of-molecular-fingerprints-in-drug-discovery-7f15021be2b1) of the fingerprinting procedure indicates that hash codes are generated at each atom before being folded into an overall molecular fingerprint.

I don't understand the question. What exactly do you want to get from the fingerprint calculation?

> Also, is it possible to compute these fingerprints from a 3D structure, rather than a SMILES string?

The RDKit can generate fingerprints for 3D molecules just the same as for "2D" molecules generated from SMILES, but almost all of the fingerprinting algorithms ignore 3D information. The one exception to this is the AtomPair fingerprinter, which you can tell to use 3D distances instead of 2D distances.

**Comment by ekwan** [comment]
Hi Greg,

Thanks for your reply.  My understanding of the fingerprinting procedure from the article is as follows.  First, atom-centered descriptors are used to generate a hash code at each atom.  Then, for each atom, data from surrounding atoms is gathered, stored in an array, and then combined into a new hash code.  These atomic hash codes are then subjected to duplicate removal and a final round of hashing to produce the molecular fingerprint.

Is it possible to retrieve the atomic hash codes before they are combined?

My question regarding 3D structures only makes sense if the answer to the above question is yes.  I want to use atomic hash codes as atomic descriptors, so it's important that I associate the atomic hash codes with the correct atom.

If the answer is no, is it possible to get the atomic descriptors for the first step so I can implement this procedure myself?  The article mentions Daylight atomic invariants, but whatever features are used in rdkit would be fine with me.

Thanks very much for your help!

-Eugene

**Reply by greglandrum** [reply]
If you don't specify which fingerprint you are interested in, then the question just can't be answered.
The RDKit has a bunch of different fingerprints available and they all treat atoms differently.

The article you link to is discussing what the RDKit calls the Morgan fingerprint. If you calculate that at radius 0 then you get just the atom hashes. The usual approaches for mapping bit information in the fingerprints back to atoms (in the RDKit documentation) can then be used to get the hashes for the atoms themselves.

**Reply by ekwan** [reply]
Hi Greg,

Thanks for your email.  The Morgan fingerprint would be fine.  When I calculated the Morgan fingerprint before, I got a bit vector.  How do I get the hashes at each atom?  If you could link to the documentation you mentioned, that would be very helpful, as I searched and couldn't find it.

Thanks so much!

-Eugene

**Reply by greglandrum** [reply]
For this operation you want to use `GetMorganFingerprint()` so that you get the unfolded fingerprints

Here's one explanation of how to get information about the bits:
https://rdkit.org/docs/GettingStartedInPython.html#explaining-bits-from-morgan-fingerprints
